pkg/bosh/manifest/dns.go

```diff
@@ -0,0 +1,323 @@
+	HeadlessServiceName(instanceGroupName string, deploymentName string) string
+
+	// Configure DNS inside pod
+	ConfigurePod(podSpec *v1.PodSpec)
+
+	// Reconcile DNS stuff
+	Reconcile(ctx context.Context, c client.Client, setOwner func(object v12.Object) error) error
+}
+
+// Target of domain alias
+type Target struct {
+	Query         string `json:"query"`
+	InstanceGroup string `json:"instance_group"`
+	Deployment    string `json:"deployment"`
+	Network       string `json:"network"`
+	Domain        string `json:"domain"`
+}
+
+// Alias of domain alias
+type Alias struct {
+	Domain  string   `json:"domain"`
+	Targets []Target `json:"targets"`
+}
+
+// boshDomainNameService is used to emulate Bosh DNS
+type boshDomainNameService struct {
+	namespace  string
+	aliases    []Alias
+	localDNSIP string
+}
+
+var _ DomainNameService = &boshDomainNameService{}
+var simple DomainNameService = &simpleDomainNameService{}
+
+// BoshDNSAddOnName name of bosh dns add on
+const BoshDNSAddOnName = "bosh-dns-aliases"
+
+// NewSimpleDomainNameService emulates old behaviour without bosh dns
+func NewSimpleDomainNameService() DomainNameService {
+	return simple
+}
+
+// NewBoshDomainNameService create a new DomainNameService
+func NewBoshDomainNameService(namespace string, addOn *AddOn) (DomainNameService, error) {
+	dns := boshDomainNameService{namespace: namespace}
+	for _, job := range addOn.Jobs {
+		aliases := job.Properties.Properties["aliases"]
+		if aliases != nil {
+			aliasesBytes, err := json.Marshal(aliases)
+			if err != nil {
+				return nil, errors.Wrapf(err, "Loading aliases from manifest")
+			}
+			var a = make([]Alias, 0)
+			err = json.Unmarshal(aliasesBytes, &a)
+			if err != nil {
+				return nil, errors.Wrapf(err, "Loading aliases from manifest")
+			}
+			dns.aliases = append(dns.aliases, a...)
+		}
+	}
+	return &dns, nil
+}
+
+// FindServiceNames see interface
+func (dns *boshDomainNameService) FindServiceNames(instanceGroupName string, deploymentName string) []string {
+	result := make([]string, 0)
+	for _, alias := range dns.aliases {
+		for _, target := range alias.Targets {
+			if target.InstanceGroup == instanceGroupName {
+				result = append(result, strings.Split(alias.Domain, ".")[0])
+			}
+		}
+	}
+	if len(result) == 0 {
+		result = append(result, serviceName(instanceGroupName, deploymentName, 63))
+	}
+	return result
+}
+
+// HeadlessServiceName see interface
+func (dns *boshDomainNameService) HeadlessServiceName(instanceGroupName string, deploymentName string) string {
+	serviceNames := dns.FindServiceNames(instanceGroupName, deploymentName)
+	if len(serviceNames) == 0 {
+		return serviceName(instanceGroupName, deploymentName, 63)
+	}
+	return serviceNames[0]
+}
+
+// ConfigurePod see interface
+func (dns *boshDomainNameService) ConfigurePod(podSpec *v1.PodSpec) {
+	if dns.localDNSIP == "" {
+		panic("BoshDomainNameService: ConfigurePod called before Reconcile")
```

Comment:
It could happen if someone calls `ConfigurePod` before `Reconcile`. 
Reading the `localIP` inside `ConfigurePod` would require passing the go k8s client through many layers. Therefore, I would like to leave it as it is.
@f0rmiga, do you have any suggestions how to handle this?

Reply:
I don't think you should change the general logic, just remove the whole check with the panic. A good comment on the function should be enough. If you still think it is very important to check, return an error instead of panicking.

Reply:
We will return an error